floodsub.go

```diff
@@ -89,12 +94,26 @@ func (fs *FloodSubRouter) Publish(from peer.ID, msg *pb.Message) {
 			continue
 		}
 
+		var success bool
 		select {
 		case mch <- out:
+			success = true
 		default:
+			success = false
 			log.Infof("dropping message to peer %s: queue full", pid)
 			// Drop it. The peer is too slow.
 		}
+
+		if hasListener {
+			select {
+			case <-listener.ctx.Done():
+			case listener.notifChan <- &msgQueuedNotification{pid, success}:
```

Comment:
This is way to heavy, it is going to spam with events for every message sent.
Also note that the channel could block the event loop, which is a no,no.

Reply:
@vyzo We could use a buffered channel here & drop the event if the buffer is full.
But, that could lead to lost events & Publish would end up erroneously reporting that it wasn't able to fulfil the target.

Reply:
So yeah, this seems quite impossible to implement correctly. 
We can't block the event loop and we don't want to erroneously report non-delivery (or not report at all).

Reply:
@raulk  Wdyt ? Can we use a buffered channel here & do the `WaitUntilQueued ` on a "best effort" basis ?

Reply:
we could launch a goroutine for every channel put, which would side-step the problem. But I worry about potentially unbounded number of goroutines running if the receiver is stalling.

Reply:
> We could use a buffered channel here & drop the event if the buffer is full.
> But, that could lead to lost events & Publish would end up erroneously reporting that it wasn't able to fulfil the target.

What we decide here is very important for the `WaitUntilQueuedALL ` as described [here](https://github.com/libp2p/go-libp2p-pubsub/issues/217#issuecomment-554855771)

Reply:
@vyzo Aha, here is one thing that can work:

1. For `WaitUntilNQueued`:
  
-  **Buffer size=N** & the event loop ONLY writes success cases to the channel
- `Topic.Publish` cancels the listener context as soon as the target is met

2. For `WaitiUntillAllQueued:
    This is slightly tricky as I've explained [here](https://github.com/libp2p/go-libp2p-pubsub/issues/217#issuecomment-554855771). But, we can do it like so:

- **Buffer size = len(pubsub.topics[topic])** when Publish is called
- Event loop writes both success & failure cases to the channel
- Event loop spins up a new go-routine if the buffer is full. Since we have estimated the buffer size based on the number of peers we know are interested in the topic when Publish is called, it effectively prevents spinning up too many go-routines.

Let me know what you think :)

Reply:
@vyzo @aschmahmann Aha, here is one thing that can work:

**For WaitUntilNQueued:**

-  `Buffer size = N` & the event loop ONLY writes success cases to the channel
- `Topic.Publish` cancels the listener context as soon as the target is met

**For WaitiUntillAllQueued:**
    This is _slightly_ tricky as I've explained [here](https://github.com/libp2p/go-libp2p-pubsub/issues/217#issuecomment-554855771). But, we can do it like so:

- `Buffer size  =  len(pubsub.topics[topic])` when Publish is called
- Event loop writes both success & failure cases to the channel
- Event loop spins up a new go-routine if the buffer is full. Since we have estimated the buffer size based on the number of peers we know are interested in the topic when Publish is called, it effectively prevents spinning up too many go-routines.

Let me know what you think :)

Reply:
Hey @aschmahmann @vyzo 

Please take a look at this when you can.